auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt

```diff
@@ -0,0 +1,173 @@
+                            + requestedAccessGroup.key,
+                )
+            }
+
+            if (!userInternal && !accessGroupData.enableOnlineRegistration) {
+                throw ApiError(
+                    HttpStatusCode.Forbidden,
+                    "external_user_non_online_registration_group",
+                    "Attempted to assign an external user a group not enabled for online registration: "
+                            + requestedAccessGroup.key,
+                )
+            }
+        }
+
+        val mapperResult = memberOfToDeltaRolesMapperFactory(
+            callingUser.cn, organisationService.findAllNamesAndCodes(), accessGroupsService.getAllAccessGroups()
+        ).map(callingUser.memberOfCNs)
+
+        val currentAccessGroups = mapperResult.accessGroups.associateBy({ it.name },
+            { it.organisationIds })
+
+        val accessGroupActions = mutableListOf<AccessGroupAction>()
+
+        val selectedOrgs = requestBodyObject.userSelectedOrgs
+
+        for (accessGroup in accessGroupsRequestMap) {
+            val currentOrganisations = currentAccessGroups[accessGroup.key]
+            val requestedOrganisations = accessGroup.value
+
+            val userIsInAccessGroup = currentOrganisations != null
+            if (userIsInAccessGroup) {
+                val organisationsAfterModification = currentOrganisations!!.toMutableSet()
+                for (requestedOrg in requestedOrganisations) {
+                    if (organisationsAfterModification.add(requestedOrg)) {
+                        accessGroupActions.add(AddAccessGroupOrganisationAction(accessGroup.key, requestedOrg))
+                    }
+                }
+                for (currentOrg in currentOrganisations) {
+                    if (!requestedOrganisations.contains(currentOrg) && selectedOrgs.contains(currentOrg)) {
+                        accessGroupActions.add(RemoveAccessGroupOrganisationAction(accessGroup.key, currentOrg))
+                        organisationsAfterModification.remove(currentOrg)
+                    }
+                }
+                if (organisationsAfterModification.isEmpty()) {
+                    accessGroupActions.add(RemoveAccessGroupAction(accessGroup.key))
+                }
+            } else {
+                if (requestedOrganisations.isNotEmpty()) {
+                    accessGroupActions.add(AddAccessGroupAction(accessGroup.key))
+                    for (requestedOrg in requestedOrganisations) {
+                        accessGroupActions.add(AddAccessGroupOrganisationAction(accessGroup.key, requestedOrg))
+                    }
+                }
+            }
+        }
+
+        for (action in accessGroupActions) {
+            if (action is AddAccessGroupAction || action is AddAccessGroupOrganisationAction) {
+                logger.atInfo().log("Adding user ${session.userCn} to access group ${action.getActiveDirectoryString()}")
```

Comment:
Should use template strings with logs, doesn't matter here, but if one of the arguments contained `{}` then it could
```suggestion
                logger.atInfo().log("Adding user {} to access group {}", session.userCn, action.getActiveDirectoryString())
```